RFD.md

```diff
@@ -0,0 +1,217 @@
+* It helps Oleksiy get a feel for what it would be like to work at Teleport, as this exercise aims to simulate their day-as-usual and expose Oleksiy to the type of work Teleport engineers are doing.
+
+## Requirements
+
+We need to build a client-server system that allows for remote execution of shell commands with per-command resource control (memory, CPU and disk IO) and isolation on network, process and filesystem levels using Linux kernel technologies like cgroups and kernel namespaces.
+
+Here are the original requirements as described on the [challenge documentation page](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge.md):
+
+
+### Library
+
+* Worker library with methods to start/stop/query status and get the output of a job.
+* Library should be able to stream the output of a running job.
+  * Output should be from start of process execution.
+  * Multiple concurrent clients should be supported.
+* Add resource control for CPU, Memory and Disk IO per job using cgroups.
+* Add resource isolation for using PID, mount, and networking namespaces.
+
+### API
+
+* [GRPC](https://grpc.io) API to start/stop/get status/stream output of a running process.
+* Use mTLS authentication and verify client certificate. Set up strong set of cipher suites for TLS and good crypto setup for certificates. Do not use any other authentication protocols on top of mTLS.
+* Use a simple authorization scheme.
+
+### Client
+
+* CLI should be able to connect to worker service and start, stop, get status, and stream output of a job.
+
+## Solution Details
+
+### Security
+
+A remote command execution service poses a number of challenges when it comes to security:
+
+- The system needs to have a strong authentication and authorization system in place to ensure that only the people allowed to perform remote commands would be able to do so.
+
+- The data coming through the system (commands, console output, etc) needs to be protected since it may contain sensitive information (credentials, customer data, etc).
+
+- The host system running the server needs to be protected from potentially abusive and/or malicious users.
+
+- Since the system is a multi-tenant environment, it may need some mechanisms to protect its users from noisy/malicious/abusive neighbors.
+
+#### Authentication
+
+We're going to use a dedicated certificate authority to issue server and client certificates and will use it to verify both the server and the client during the establishment of an mTLS connection to the service.
+
+##### Scope limits
+
+CA, server and client certificate generation and management is out of scope for the project. A set of example certificates and keys will be checked into the repository with the project to allow for easier development of the system.
```

Comment:
How do you plan to configure TLS + algorithm for certs?

Reply:
Updated the RFD with a clarified set of details around the TLS algorithm and ciphers to be used by the solution.